cmd/mpi-operator.v1alpha2/app/server.go

```diff
@@ -67,49 +93,138 @@ func Run(opt *options.ServerOption) error {
 		glog.Fatalf("Error building kubeConfig: %s", err.Error())
 	}
 
-	kubeClient, err := kubernetes.NewForConfig(cfg)
+	// Create clients.
+	kubeClient, leaderElectionClientSet, mpiJobClientSet, err := createClientSets(cfg)
 	if err != nil {
-		glog.Fatalf("Error building kubernetes clientset: %s", err.Error())
+		return err
+	}
+	if !checkCRDExists(mpiJobClientSet, opt.Namespace) {
+		glog.Info("CRD doesn't exist. Exiting")
+		os.Exit(1)
 	}
 
-	kubeflowClient, err := clientset.NewForConfig(cfg)
+	// Set leader election start function.
+	run := func(ctx context.Context) {
+		var kubeInformerFactory kubeinformers.SharedInformerFactory
+		var kubeflowInformerFactory informers.SharedInformerFactory
+		if opt.Namespace == "" {
+			kubeInformerFactory = kubeinformers.NewSharedInformerFactory(kubeClient, 0)
+			kubeflowInformerFactory = informers.NewSharedInformerFactory(mpiJobClientSet, 0)
+		} else {
+			kubeInformerFactory = kubeinformers.NewSharedInformerFactoryWithOptions(kubeClient, 0, kubeinformers.WithNamespace(opt.Namespace), nil)
+			kubeflowInformerFactory = informers.NewSharedInformerFactoryWithOptions(mpiJobClientSet, 0, informers.WithNamespace(opt.Namespace), nil)
+		}
+
+		var pdbInformer policyinformers.PodDisruptionBudgetInformer
+		if opt.EnableGangScheduling {
+			pdbInformer = kubeInformerFactory.Policy().V1beta1().PodDisruptionBudgets()
+		}
+		controller := controllersv1alpha2.NewMPIJobController(
+			kubeClient,
+			mpiJobClientSet,
+			kubeInformerFactory.Core().V1().ConfigMaps(),
+			kubeInformerFactory.Core().V1().ServiceAccounts(),
+			kubeInformerFactory.Rbac().V1().Roles(),
+			kubeInformerFactory.Rbac().V1().RoleBindings(),
+			kubeInformerFactory.Apps().V1().StatefulSets(),
+			kubeInformerFactory.Batch().V1().Jobs(),
+			pdbInformer,
+			kubeflowInformerFactory.Kubeflow().V1alpha2().MPIJobs(),
+			opt.KubectlDeliveryImage,
+			opt.EnableGangScheduling)
+
+		go kubeInformerFactory.Start(ctx.Done())
+		go kubeflowInformerFactory.Start(ctx.Done())
+
+		if err = controller.Run(opt.Threadiness, stopCh); err != nil {
+			glog.Fatalf("Error running controller: %s", err.Error())
+		}
+	}
+
+	id, err := os.Hostname()
 	if err != nil {
-		glog.Fatalf("Error building kubeflow clientset: %s", err.Error())
+		return fmt.Errorf("failed to get hostname: %v", err)
+	}
+	// add a uniquifier so that two processes on the same host don't accidentally both become active
+	id = id + "_" + string(uuid.NewUUID())
```

Comment:
if don't add uuid to id, when the host has same hostname, the process will all become active.